### PR TITLE
Add shuffle reduction for MDRange

### PR DESCRIPTION
**Pull request by Rombur**
The code is very similar to the `Range` code.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp [hunk 1]
```diff
@@ -227,23 +224,29 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
+  // when to use shuffle
+  static bool constexpr UseShflReduction =
+      (sizeof(value_type) > 2 * sizeof(double)) &&
+      (ValueTraits::StaticValueSize != 0);
+
  private:
-  using DummyShflReductionType  = double;
-  using DummySHMEMReductionType = int;
+  struct ShflReductionTag {};
+  struct SHMEMReductionTag {};
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:235]:
```suggestion
  struct ShmemReductionTag {};
```
I know it's pre-existent but this capitalization still weird. We can address that in a different pull request, though.

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp [hunk 2]
```diff
@@ -227,23 +224,29 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
   }
 
   inline __device__ void operator()(void) const {
+    using ReductionTag =
+        typename std::conditional<UseShflReduction, ShflReductionTag,
+                                  SHMEMReductionTag>::type;
+    run(ReductionTag{});
+  }
+
+  __device__ inline void run(SHMEMReductionTag) const {
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:249]: We can just remove all the `inline` specifiers.

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp [hunk 3]
```diff
@@ -296,6 +299,44 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
 
+  __device__ inline void run(ShflReductionTag) const {
+    value_type value;
+    ValueInit::init(ReducerConditional::select(m_functor, m_reducer), &value);
+    // Number of blocks is bounded so that the reduction can be limited to two
+    // passes. Each thread block is given an approximately equal amount of work
+    // to  perform. Accumulate the values for this block. The accumulation
+    // ordering does not match the final pass, but is arithmetically equivalent.
+
+    const Member work_part =
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:310]: You only use this variable in l.318. Can you move it there?

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp [hunk 4]
```diff
@@ -296,6 +299,44 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
+    max_active_thread =
+        (max_active_thread == 0) ? blockDim.y : max_active_thread;
+
+    value_type init;
+    ValueInit::init(ReducerConditional::select(m_functor, m_reducer), &init);
+    if (Impl::hip_inter_block_reduction<ReducerTypeFwd, ValueJoin, WorkTagFwd>(
+            value, init,
+            ValueJoin(ReducerConditional::select(m_functor, m_reducer)),
+            m_scratch_space, result, m_scratch_flags, max_active_thread)) {
+      const unsigned id = threadIdx.y * blockDim.x + threadIdx.x;
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:328]:
```suggestion
      unsigned const id = threadIdx.y * blockDim.x + threadIdx.x;
```

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp (hunk 1 above)
Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:235]: What about `ShMemReductionTag` since it's really Shared Memory?

**Comment by Rombur** [comment]
This PR is based on the commented code from CUDA but the CUDA code never worked. So I need to rethink this.

File: core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp (hunk 4 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp:328]: No that is not required. I like const unsigned much more :-)

**Pull request closed without merging**